src/Security/Authenticator.php

```diff
@@ -117,9 +117,10 @@ public static function register_authenticator($authenticator)
             return false;
         }
 
-        if (in_array($authenticator, self::$authenticators) == false) {
-            if (call_user_func(array($authenticator, 'on_register')) === true) {
-                array_push(self::$authenticators, $authenticator);
+        $authenticators = self::config()->get('authenticators');
+        if (in_array($authenticator, $authenticators) == false) {
+            if ($authenticator::on_register() === true) {
+                self::config()->merge('authenticators', [$authenticator]);
```

Comment:
Any time that configs are modified, you immediately lose a level of performance, since the immutable config collection needs to be cloned before update.

Config should not be used as a cache; Can we put this somewhere else?